nifi-nar-bundles/nifi-standard-bundle/nifi-standard-processors/src/main/java/org/apache/nifi/processors/standard/DeduplicateRecords.java

```diff
@@ -0,0 +1,690 @@
+import java.util.HashSet;
+import java.util.List;
+import java.util.Map;
+import java.util.Set;
+import java.util.stream.Collectors;
+
+import static java.util.stream.Collectors.toList;
+import static org.apache.commons.codec.binary.StringUtils.getBytesUtf8;
+
+@EventDriven
+@SupportsBatching
+@InputRequirement(Requirement.INPUT_REQUIRED)
+@SystemResourceConsideration(resource = SystemResource.MEMORY,
+    description = "The HashSet filter type will grow memory space proportionate to the number of unique records processed. " +
+        "The BloomFilter type will use constant memory regardless of the number of records processed.")
+@SystemResourceConsideration(resource = SystemResource.CPU,
+    description =  "If a more advanced hash algorithm is chosen, the amount of time required to hash any particular " +
+            "record could increase substantially."
+)
+@Tags({"text", "record", "update", "change", "replace", "modify", "distinct", "unique",
+    "filter", "hash", "dupe", "duplicate", "dedupe"})
+@CapabilityDescription("This processor attempts to deduplicate a record set in memory using either a hashset or a bloom filter. " +
+        "It operates on a per-file basis rather than across an entire data set that spans multiple files.")
+@WritesAttribute(attribute = "record.count", description = "The number of records processed.")
+@DynamicProperty(
+    name = "RecordPath",
+    value = "An expression language statement used to determine how the RecordPath is resolved. " +
+            "The following variables are availible: ${field.name}, ${field.value}, ${field.type}",
+    description = "The name of each user-defined property must be a valid RecordPath.")
+@SeeAlso(classNames = {
+    "org.apache.nifi.distributed.cache.client.DistributedMapCacheClientService",
+    "org.apache.nifi.distributed.cache.server.map.DistributedMapCacheServer",
+    "org.apache.nifi.processors.standard.DetectDuplicate"
+})
+public class DeduplicateRecords extends AbstractProcessor {
+
+    private static final String FIELD_NAME = "field.name";
+    private static final String FIELD_VALUE = "field.value";
+    private static final String FIELD_TYPE = "field.type";
+
+    private volatile RecordPathCache recordPathCache;
+    private volatile List<String> recordPaths;
+
+    // VALUES
+
+    static final AllowableValue NONE_ALGORITHM_VALUE = new AllowableValue("none", "None",
+            "Do not use a hashing algorithm. The value of resolved RecordPaths will be combined with tildes (~) to form the unique record key. " +
+                    "This may use significantly more storage depending on the size and shape or your data.");
+    static final AllowableValue MD5_ALGORITHM_VALUE = new AllowableValue(MessageDigestAlgorithms.MD5, "MD5",
+            "The MD5 message-digest algorithm.");
+    static final AllowableValue SHA1_ALGORITHM_VALUE = new AllowableValue(MessageDigestAlgorithms.SHA_1, "SHA-1",
+            "The SHA-1 cryptographic hash algorithm.");
```

Comment:
Since some time has past since the original submission, recommend removing the `MD5` and `SHA-1` algorithms.  Given known possibilities for hash collisions, and the minimal cost of `SHA-256`, there does not seem to be any value in supporting these options.  It is worth noting that NiFi itself switched from MD5 to SHA-256 for NAR digest calculation over a year ago.